src/sentry/rules/processing/buffer_processing.py

```diff
@@ -0,0 +1,155 @@
+
+from sentry import buffer, options
+from sentry.buffer.redis import BufferHookEvent, redis_buffer_registry
+from sentry.db import models
+from sentry.rules.processing.processor import PROJECT_ID_BUFFER_LIST_KEY
+from sentry.utils import metrics
+from sentry.utils.registry import NoRegistrationExistsError, Registry
+
+logger = logging.getLogger("sentry.delayed_processing")
+
+
+@dataclass
+class FilterKeys:
+    project_id: int
+
+
+@dataclass
+class BufferHashKeys:
+    model: type[models.Model]
+    filters: FilterKeys
+
+
+class DelayedProcessingBase(ABC):
+    def __init__(self, project_id: int):
+        self.project_id = project_id
+
+    @property
+    @abstractmethod
+    def hash_args(self) -> BufferHashKeys:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def processing_task(self) -> Task:
+        raise NotImplementedError
+
+
+delayed_processing_registry = Registry[type[DelayedProcessingBase]]()
+
+
+def fetch_alertgroup_to_event_data(
+    project_id: int, model: type[models.Model], batch_key: str | None = None
+) -> dict[str, str]:
+    field: dict[str, models.Model | int | str] = {
+        "project_id": project_id,
+    }
+
+    if batch_key:
+        field["batch_key"] = batch_key
+
+    return buffer.backend.get_hash(model=model, field=field)
+
+
+def bucket_num_groups(num_groups: int) -> str:
+    if num_groups > 1:
+        magnitude = 10 ** int(math.log10(num_groups))
+        return f">{magnitude}"
+    return "1"
+
+
+def process_project_alerts_in_batches(project_id: int, processing_type: str) -> None:
+    """
+    This will check the number of alertgroup_to_event_data items in the Redis buffer for a project.
+
+    If the number is larger than the batch size, it will chunk the items and process them in batches.
+
+    The batches are replicated into a new redis hash with a unique filter (a uuid) to identify the batch.
+    We need to use a UUID because these batches can be created in multiple processes and we need to ensure
+    uniqueness across all of them for the centralized redis buffer. The batches are stored in redis because
+    we shouldn't pass objects that need to be pickled and 10k items could be problematic in the celery tasks
+    as arguments could be problematic. Finally, we can't use a pagination system on the data because
+    redis doesn't maintain the sort order of the hash keys.
+
+    `processing_task` will fetch the batch from redis and process the rules.
+    """
+    batch_size = options.get("delayed_processing.batch_size")
+    log_format = "{}.{}"
+
+    try:
+        processing_info = delayed_processing_registry.get(processing_type)(project_id)
+    except NoRegistrationExistsError:
+        logger.exception(log_format.format(processing_type, "no_registration"))
+        return
+
+    hash_args = processing_info.hash_args
+    task = processing_info.processing_task
+    filters: dict[str, models.Model | str | int] = asdict(hash_args.filters)
```

Comment:
this is new, depending on the `processing_type` (basically if we're processing rules or workflows) we'll have different args to pass when getting things from the buffer or pushing to it, and we'll have a different task to kick off to query snuba to evaluate slow conditions.

the rest of the logic below just uses this information and is the same as before.

Reply:
could we give `models.Model | str | int` type a name to help illustrate the different args for this?